automated install/basic-install.sh

```diff
@@ -19,7 +19,8 @@
 # -e option instructs bash to immediately exit if any command [1] has a non-zero exit status
 # We do not want users to end up with a partially working install, so we exit the script
 # instead of continuing the installation with something broken
-set -e
+# -e is now turned off. All important commands should be invoked using run_command.
```

Comment:
`set -e` catches the unknown, rather than just what we wrap in a defensive execute function.  I think it will always be a best practice to enable set -e rather than think you can defensively code against every possible error.  The risk of shutting it off is a user thinks the script ran successfully, when in fact it did not, and the actual error gets buried underneath a bunch of output that followed the erorr.

Reply:
I've gone back and forth on the `set -e` debate, if the script is set to instead trap everything and check on exit for any errors, then I'm okay with it not being enabled. But just disabling `set -e` without a proper plan of attack for doing things is something we'll need to discuss. If you want to just `set +e` at the top of `run_command` and then reset `set -e` before `run_command` exits, then that may be an approach as well.

Reply:
I don't think something like wrapping each `install -m755` call in a separate `run_command` would be useful, and I do consider every command and function to be important. I'd like to not have every other line be a wrapped call to a function.

Reply:
Doing `set -e` at the start of the script and removing `run_command` from basic Unix/Linux commands  (`mkdir, cp, install`) that create output only when something goes wrong would work, *so long as output from those commands is not redirected to /dev/null.* Then `run_command` would be needed only on the few commands and programs that always send something to stdout, such as `${PKG_INSTALL}` and `systemctl.`

This way the goal of this PR will still be accomplished: output from `basic-install.sh` will be as clean as it is now, but the user will always have some output if a command fails. And it means we can remove the `run_command` wrapper from about 90% of the lines where I have it now.

For now I'll fix up the `\n` that stickler is complaining about.

Reply:
For the common utility commands like `mkdir, cp, install` it would be something to consider with placing them in an `if` construct. Still output to `/dev/null` but check the retval of the utility and if it fails we can present our own message. I'm good with something like `install -m755 <file> || { echo "This failed"; exit 1 }` kind of thing, with a better echo message and either an `exit` or a `return` depending on the severity of the error and the usage of the function.

Reply:
Eventually I'd go with something more elegant and trap things, so if there is a failure we can handle it and present the user with the requisite message and behaviour. We just went with `set -e` as a default to try and catch what we could, but I do like the idea of traps as more functional. I have a lot of code to come in with changes to the installation script but I'm holding on that so I don't throw a bunch of modifications in to the mix while we are working on this PR. The installer will have some major changes, but I'd like to include `run_command` as part of those changes, I like the concept and the execution and it just makes sense.

Reply:
I'm not a fan of throwing away output when something fails. That leads to a situation where the message is, in effect, "Something went wrong. I can't tell you why, so you'll have to guess." It's too reminiscent of messages from a popular operating system: "Operation failed. [OK] [Cancel]."

Reply:
Interesting conumdrum. Being able to present useful error messages when a command fails--including the text of the failed command and its output--is the goal of `run_command.` But for full coverage that means having to wrap even straightforward commands like `mkdir` and `cp`, making for (currently) 88 calls to `run_command. That's because using `set -e` while not redirecting stderr makes for non-constructive error messages, and the command that failed isn't displayed to the user. 

In lieu of `set -e`, using `trap err_handler ERR` (which uses the same semantics as `set -e`) looks promising. But we're back to non-constructive error messages because the command's output is displayed before the trap handler kicks in. In this case sending all output to a file descriptor might work ... but write descriptors are appending because they're opened only once, making it difficult to determine where the output from the command that triggered the trap starts. To get around that problem, one could tag the output descriptor every time a command is issued (assuming there's a common output file opened on descriptor 4):

    # Set up an error handler for failed commands
    trap err_trap ERR
    err_trap() {
        ("an error occurred")
        (display command that failed, available in $BASH_COMMAND)
        (display output from FD 4 starting from __TAG_${TAG_COUNTER}__
        (display call trace)
        (exit unless error is non-fatal)
    }
    # TAG_COUNTER is used by trap_err to know where output from the trapped command starts
    TAG_COUNTER=0
    # Add a tag the stdout/stderr file descriptor just before running a command
    output_tag() { echo "__TAG_$((++TAG_COUNTER))__" >&4 }
    ....
    # Set up a file to contain output from various command and assign it file descriptor 4
    stdOutErrFile="$(mktemp pihole_stdouterr_XXXX)"
    exec 4>"$stdOutErrFile"
    rm -f "$stdOutErrFile"
    ....
    output_tag; cp -a from_file target &>4

It's almost a complete replacement for `run_command`, and aside from having to run `output_tag` prior to a command that sends stdout and stderr to `&4` it's cleaner then wrapping nearly a hundred commands with `run_command.` And in the call to `output_tag` one could differentiate between fatal (default) and non-fatal errors.

One issue with this is a trapped error could show the wrong information because `output_tag` wasn't called and/or output wasn't directed to a file descriptor.

Reply:
That's great information, thanks. I think it may just be a convention that we adopt if we need to. The other options are going to be appending `||` to the calls `cp -a from_file target || echo "command specific failure message"` So prepending with a guardian function may be more functional.

Reply:
Is `run_command` still something that is going to continue, or is the trap functionality a better approach? I have a few PRs that I'd like to get going and if `run_command` is going to be modified it would be better for that function to be based against the results of the new PRs.

Reply:
The trap functionality is better. Let's close this PR. Later today or tomorrow I'll open a new one with the trap handler.

Reply:
Thanks, looking forward to the new PR.